cmd/object-handlers.go

```diff
@@ -85,11 +81,61 @@ func (api objectAPIHandlers) SelectObjectContentHandler(w http.ResponseWriter, r
 		return
 	}
 
-	getObjectInfo := objectAPI.GetObjectInfo
+	if crypto.S3.IsRequested(r.Header) || crypto.S3KMS.IsRequested(r.Header) { // If SSE-S3 or SSE-KMS present -> AWS fails with undefined error
+		writeErrorResponse(w, ErrBadRequest, r.URL)
+		return
+	}
+
+	vars := mux.Vars(r)
+	bucket := vars["bucket"]
+	object := vars["object"]
+
+	// Get request range.
+	rangeHeader := r.Header.Get("Range")
+	if rangeHeader != "" {
+		writeErrorResponse(w, ErrUnsupportedRangeHeader, r.URL)
+		return
+	}
+
+	if r.ContentLength <= 0 {
+		writeErrorResponse(w, ErrEmptyRequestBody, r.URL)
+		return
+	}
+
+	var selectReq ObjectSelectRequest
+	if err := xmlDecoder(r.Body, &selectReq, r.ContentLength); err != nil {
+		writeErrorResponse(w, ErrMalformedXML, r.URL)
+		return
+	}
+
+	if !strings.EqualFold(string(selectReq.ExpressionType), "SQL") {
+		writeErrorResponse(w, ErrInvalidExpressionType, r.URL)
+		return
+	}
+	if len(selectReq.Expression) >= s3select.MaxExpressionLength {
+		writeErrorResponse(w, ErrExpressionTooLong, r.URL)
+		return
+	}
+	if selectReq.InputSerialization.CSV.FileHeaderInfo != CSVFileHeaderInfoUse &&
+		selectReq.InputSerialization.CSV.FileHeaderInfo != CSVFileHeaderInfoNone &&
+		selectReq.InputSerialization.CSV.FileHeaderInfo != CSVFileHeaderInfoIgnore &&
+		selectReq.InputSerialization.CSV.FileHeaderInfo != "" {
+		writeErrorResponse(w, ErrInvalidFileHeaderInfo, r.URL)
+		return
+	}
+	if selectReq.OutputSerialization.CSV.QuoteFields != CSVQuoteFieldsAlways &&
+		selectReq.OutputSerialization.CSV.QuoteFields != CSVQuoteFieldsAsNeeded &&
+		selectReq.OutputSerialization.CSV.QuoteFields != "" {
+		writeErrorResponse(w, ErrInvalidQuoteFields, r.URL)
+		return
+	}
+
+	getObjectNInfo := objectAPI.GetObjectNInfo
 	if api.CacheAPI() != nil {
-		getObjectInfo = api.CacheAPI().GetObjectInfo
+		getObjectNInfo = api.CacheAPI().GetObjectNInfo
 	}
 
+	objInfo, reader, err := getObjectNInfo(ctx, bucket, object, nil)
```

Comment:
Why do we call `getObjectNInfo` before the request authentication was checked - this seems unnecessary and would lead to fetching the object twice in case of e.g. encryption...  See https://github.com/minio/minio/pull/6370/files#diff-d1c4dccc78d8d9f128ca8972a1199176R201

Reply:
We need to otherwise we cannot send an S3 compatible error underneath.

Reply:
We don't such complication we need to make sure that getObjectInfo and getObject are not disconnected operations. But just for Auth it can be used since these get called only for anonymous operations.

Reply:
Not able to understand... My point is that the first call to `GetObjectNInfo` (before auth check) is not necessary in all cases where the auth-check is passed **and** https://github.com/minio/minio/pull/6370/files#diff-d1c4dccc78d8d9f128ca8972a1199176R200 is hit. Then we call `GetObjectNInfo` twice...
However this should be an edge case - so probably doesn't matter much...

Reply:
of course we need, we used to do GetObjectInfo() we simply do GetObjectNInfo() now. But I changed this in the separate PR where we rely on GetObjectInfo() instead.